src/memory_pool.c

```diff
@@ -58,9 +58,15 @@ size_t umfPoolMallocUsableSize(umf_memory_pool_handle_t hPool, void *ptr) {
     return hPool->ops.malloc_usable_size(hPool->pool_priv, ptr);
 }
 
+umf_result_t umfPoolFreeSized(umf_memory_pool_handle_t hPool, void *ptr,
+                              size_t size) {
+    UMF_CHECK((hPool != NULL), UMF_RESULT_ERROR_INVALID_ARGUMENT);
+    return hPool->ops.free(hPool->pool_priv, ptr, size);
+}
+
 umf_result_t umfPoolFree(umf_memory_pool_handle_t hPool, void *ptr) {
     UMF_CHECK((hPool != NULL), UMF_RESULT_ERROR_INVALID_ARGUMENT);
-    return hPool->ops.free(hPool->pool_priv, ptr);
+    return hPool->ops.free(hPool->pool_priv, ptr, 0);
```

Comment:
As this is non-default API I think this should not be required to have this function in ops. So I would not change the regular free signature, but add an extra function to the ops. Note that the optimizations you mentioned have to be implemented in pools, so it makes sense to the pool to expose this function only if using it makes sense. Also, as a fallback (if there is no such function in the ops) we could just call free(ptr) and ignore the size